lib/http/node.js

```diff
@@ -21,112 +22,107 @@ const Logger = require('../logger');
  * @return  {Number} number of milliseconds to wait before retrying again the request.
  */
 NodeHttpClient.prototype.constructExponentialBackoffStrategy = function () {
-  var sleep = this._connectionConfig.getRetrySfStartingSleepTime();
-  var cap = this._connectionConfig.getRetrySfMaxSleepTime();
-  var base = 1;
+  let sleep = this._connectionConfig.getRetrySfStartingSleepTime();
+  const cap = this._connectionConfig.getRetrySfMaxSleepTime();
+  const base = 1;
   sleep = Util.nextSleepTime(base, cap, sleep);
   return sleep * 1000;
-}
+};
 
 /**
  * Creates a client that can be used to make requests in Node.js.
  *
  * @param {ConnectionConfig} connectionConfig
  * @constructor
  */
-function NodeHttpClient(connectionConfig)
-{
+function NodeHttpClient(connectionConfig) {
   Base.apply(this, arguments);
 }
 
 Util.inherits(NodeHttpClient, Base);
 
-/**
- * @inheritDoc
- */
-NodeHttpClient.prototype.getAgentAndProxyOptions = function (url, proxy, mock)
-{
-  const isHttps = Url.parse(url).protocol === 'https:';
-  let agentClass;
-  let agentOptions = { keepAlive: true };
-  let options;
-  var bypassProxy = false;
+const httpsAgentCache = new Map();
+
+function getFromCacheOrCreate(agentClass, options, parsedUrl) {
+  const protocol = parsedUrl.protocol;
+  const port = parsedUrl.port || (protocol === 'http:' ? '80' : '443');
+  const agentId = `${protocol}//${parsedUrl.hostname}:${port}`;
 
-  if (proxy && proxy.noProxy)
-  {
-    var bypassList = proxy.noProxy.split("|");
-    for (let i = 0; i < bypassList.length; i++)
-    {
+  let agent ={}
+  function createAgent(agentClass, agentOptions) {
+    const agent = agentClass(agentOptions);
+    httpsAgentCache.set(agentId, agent);
+    Logger.getInstance().trace(`Create and add to cache new agent ${agentId}`);
+    return agent;
+  }
+
+  if (httpsAgentCache.has(agentId)) {
+    Logger.getInstance().trace(`Get agent with id: ${agentId} from cache`);
+    agent = httpsAgentCache.get(agentId);
+    // recreate agent if keepAlive parameter was changed
+    if (agent.keepAlive !== options.keepAlive) {
```

Comment:
cannot keepAlive be a part of the cache key and then instead of creating new agent we are just reusing agent with turned off keepAlive?

Reply:
It was first idea ... maybe first is the best ;)